partiql-logical-planner/src/lib.rs

```diff
@@ -0,0 +1,1397 @@
+    }
+
+    fn exit_join(&mut self, _join: &'ast Join) {
+        let mut benv = self.exit_benv();
+        assert_eq!(benv.len(), 2);
+
+        let mut env = self.exit_env();
+        assert!((0..1).contains(&env.len()));
+
+        let Join { kind, .. } = _join;
+
+        let kind = match kind {
+            JoinKind::Inner => logical::JoinKind::Inner,
+            JoinKind::Left => logical::JoinKind::Left,
+            JoinKind::Right => logical::JoinKind::Right,
+            JoinKind::Full => logical::JoinKind::Full,
+            JoinKind::Cross => logical::JoinKind::Cross,
+        };
+
+        let on = env.pop();
+
+        let rid = benv.pop().unwrap();
+        let lid = benv.pop().unwrap();
+        let left = Box::new(self.plan.operator(lid).unwrap().clone());
+        let right = Box::new(self.plan.operator(rid).unwrap().clone());
+        let join = logical::BindingsOp::Join(logical::Join {
+            kind,
+            on,
+            left,
+            right,
+        });
+        let join = self.plan.add_operator(join);
+        self.push_bexpr(join);
+    }
+
+    fn enter_join_spec(&mut self, _join_spec: &'ast JoinSpec) {
+        match _join_spec {
+            JoinSpec::On(_) => {
+                // visitor recurse into expr will put the condition in the current env
+            }
+            JoinSpec::Using(_) => {
+                todo!("JoinSpec::Using")
+            }
+            JoinSpec::Natural => {
+                todo!("JoinSpec::Natural")
+            }
+        };
+    }
+
+    fn exit_join_spec(&mut self, _join_spec: &'ast JoinSpec) {}
+
+    fn enter_where_clause(&mut self, _where_clause: &'ast ast::WhereClause) {
+        self.enter_env();
+    }
+
+    fn exit_where_clause(&mut self, _where_clause: &'ast ast::WhereClause) {
+        let mut env = self.exit_env();
+        assert_eq!(env.len(), 1);
+
+        let filter = logical::BindingsOp::Filter(logical::Filter {
+            expr: env.pop().unwrap(),
+        });
+        let id = self.plan.add_operator(filter);
+
+        self.current_clauses_mut().where_clause.replace(id);
+    }
+
+    fn enter_having_clause(&mut self, _having_clause: &'ast ast::HavingClause) {
+        self.enter_env();
+    }
+
+    fn exit_having_clause(&mut self, _having_clause: &'ast ast::HavingClause) {
+        let _env = self.exit_env();
+        todo!("having clause");
+    }
+
+    fn enter_group_by_expr(&mut self, _group_by_expr: &'ast GroupByExpr) {
+        self.enter_env();
+    }
+
+    fn exit_group_by_expr(&mut self, _group_by_expr: &'ast GroupByExpr) {
+        let _env = self.exit_env();
+        todo!("group by clause");
+    }
+
+    fn enter_order_by_expr(&mut self, _order_by_expr: &'ast OrderByExpr) {
+        self.enter_env();
+    }
+
+    fn exit_order_by_expr(&mut self, _order_by_expr: &'ast OrderByExpr) {
+        let _env = self.exit_env();
+        todo!("order by clause");
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
```

Comment:
Could include some further documentation on how the name resolution and AST to Plan compilation works. Not sure if we want to include documentation before or after the 0.2.0 release, but having some supporting documentation could help with understanding the complicated parts of PartiQL's name resolution and how it's implemented.